### PR TITLE
Base64 decode: handle invalid inputs

### PR DESCRIPTION
**Pull request by mtakigiku**
Fixed the following problems of `Base64::decode()`:
- It said it skips invalid characters, but no reason to do that. It should treat them as invalid input and return empty string.
- There should not be unused bits but it was not handled. (e.g. `"AB=="` (`000000 000001`) should be an invalid input because the last 4 bits of second character `B` contains `1`)

### CODE REVIEW

**Comment by mattklein123** [comment]
Reviewing these kind of PRs gives me a headache. @RomanDzhabarov @PiotrSikora @dnoe I feel that all of you probably enjoy this kind of thing so please review. :)

**Comment by RomanDzhabarov** [comment]
Yup, will do over the weekend, this is fun!

File: source/common/common/base64.cc [hunk 1]
```diff
@@ -49,23 +53,47 @@ std::string Base64::decode(const std::string& input) {
     // Take first 6 bits from 1st converted char and first 2 bits from 2nd converted char,
     // make 8 bits char from it.
     // Use conversion table to map char to decoded value (value is between 0 and 64).
-    result.push_back(REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read])] << 2 |
-                     REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 1])] >> 4);
+    unsigned char a = REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read])];
+    unsigned char b = REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 1])];
+    if (a == 64 || b == 64) {
+      // input contains an invalid character.
```
Comment (RomanDzhabarov) [inline source/common/common/base64.cc]: nit: s/input/Input

File: test/common/common/base64_test.cc [hunk 2]
```diff
@@ -74,6 +75,16 @@ TEST(Base64Test, Decode) {
...
 }
 
+TEST(Base64Test, DecodeFailure) {
+  EXPECT_EQ("", Base64::decode("==Zg"));
+  EXPECT_EQ("", Base64::decode("=Zm8"));
+  EXPECT_EQ("", Base64::decode("Zm=8"));
+  EXPECT_EQ("", Base64::decode("Zg=A"));
+  EXPECT_EQ("", Base64::decode("Zm9=")); // 011001 100110 111101 <- unused bit at tail
+  EXPECT_EQ("", Base64::decode("Zg.."));
+  EXPECT_EQ("", Base64::decode("..Zg"));
```
Comment (RomanDzhabarov) [inline test/common/common/base64_test.cc]: I'd also add a test with the string ending by ===

File: source/common/common/base64.cc [hunk 3]
```diff
@@ -49,23 +53,47 @@ std::string Base64::decode(const std::string& input) {
     // Take first 6 bits from 1st converted char and first 2 bits from 2nd converted char,
     // make 8 bits char from it.
     // Use conversion table to map char to decoded value (value is between 0 and 64).
-    result.push_back(REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read])] << 2 |
-                     REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 1])] >> 4);
+    unsigned char a = REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read])];
```
Comment (RomanDzhabarov) [inline source/common/common/base64.cc]: nit: could be `const unsigned char a`

File: source/common/common/base64.cc [hunk 4]
```diff
@@ -49,23 +53,47 @@ std::string Base64::decode(const std::string& input) {
     // Take first 6 bits from 1st converted char and first 2 bits from 2nd converted char,
     // make 8 bits char from it.
     // Use conversion table to map char to decoded value (value is between 0 and 64).
```
Comment (dnoe) [inline source/common/common/base64.cc]: This comment is now (I suppose it always was) slightly misleading, the value is between 0 and 63 inclusive if all is good. Can you update the comment?

File: source/common/common/base64.cc [hunk 5]
```diff
@@ -49,23 +53,47 @@ std::string Base64::decode(const std::string& input) {
...
-    result.push_back(REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read])] << 2 |
-                     REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 1])] >> 4);
+    unsigned char a = REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read])];
+    unsigned char b = REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 1])];
+    if (a == 64 || b == 64) {
+      // input contains an invalid character.
+      return EMPTY_STRING;
+    }
+    result.push_back(a << 2 | b >> 4);
     unsigned char c = REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 2])];
```
Comment (dnoe) [inline source/common/common/base64.cc]: nit: make this one const too while you're there.

File: source/common/common/base64.cc [hunk 6]
```diff
@@ -49,23 +53,47 @@ std::string Base64::decode(const std::string& input) {
...
-      result.push_back(REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 1])] << 4 |
-                       c >> 2);
-      unsigned char d = REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 3])];
-
-      if (d < 64) {
-        // Take last 2 bits from 3rd converted char and all(6) bits from 4th converted char.
-        result.push_back(REVERSE_LOOKUP_TABLE[static_cast<uint32_t>(input[cur_read + 2])] << 6 | d);
+    // Decoded value 64 means invalid character unless we already know it is a valid padding.
+    // If so, following characters are all padding.
+    // Also we should check there is no unused bits.
```
Comment (dnoe) [inline source/common/common/base64.cc]: nit: "there are no unused bits"

File: test/common/common/base64_test.cc [hunk 7]
```diff
@@ -74,6 +75,16 @@ TEST(Base64Test, Decode) {
   }
 }
 
+TEST(Base64Test, DecodeFailure) {
+  EXPECT_EQ("", Base64::decode("==Zg"));
+  EXPECT_EQ("", Base64::decode("=Zm8"));
+  EXPECT_EQ("", Base64::decode("Zm=8"));
+  EXPECT_EQ("", Base64::decode("Zg=A"));
+  EXPECT_EQ("", Base64::decode("Zm9=")); // 011001 100110 111101 <- unused bit at tail
```
Comment (dnoe) [inline test/common/common/base64_test.cc]: Does this one test case cover both of the unused bit at tail cases above? Seems like there should be two but maybe I'm missing something.

File: test/common/common/base64_test.cc [hunk 8]
```diff
@@ -39,6 +39,7 @@ TEST(Base64Test, Decode) {
   EXPECT_EQ("", Base64::decode(""));
   EXPECT_EQ("foo", Base64::decode("Zm9v"));
   EXPECT_EQ("fo", Base64::decode("Zm8="));
+  EXPECT_EQ("f", Base64::decode("Zg=="));
   EXPECT_EQ("foobar", Base64::decode("Zm9vYmFy"));
   EXPECT_EQ("foob", Base64::decode("Zm9vYg=="));
   EXPECT_EQ("", Base64::decode("123"));
```
Comment (dnoe) [inline test/common/common/base64_test.cc]: I think "123" is a failure case, can you move it below into the failure test?

File: source/common/common/base64.cc (hunk 4 above)
Comment (RomanDzhabarov) [inline source/common/common/base64.cc]: @dnoe you can blame me :) yeah, let's fix it now.

File: test/common/common/base64_test.cc (hunk 7 above)
Comment (mtakigiku) [inline test/common/common/base64_test.cc]: You're right. I've just added a test:)

**Comment by mtakigiku** [comment]
Thank you everyone for your comments. I've just updated it :)

**Pull request merged**